feat: Update keystore structure (#259)
awslabs/aws-encryption-sdk-specification: changes/2023_7_12_update-keystore-structure/background.md

Since the hierarchy has a cache, this problem may not even be small.

It is significantly better for the request to be as atomic as possible.

## How do we distinguish active from non-active versions?

### _Insert 2 records, the active record and the version record_

Here we insert the version record and that data is immutable.
Then we also insert a “version” that is the active version.
Having an immutable record is a nice property.
The downside here is that there are 2 records to insert.
The other advantage here is that getting this record by version is easy.
This is important because on decrypt we get the branch key by version and not by “active”.

### ReEncrypt the active version to no longer be active

Here we would insert a single record and then reEncrypt this value to be no longer active.
This means that the transaction to update is more complicated.
I think that it is still possible to update everything, but now there is a single copy of the branch key that is getting updated.
It seems impossible, but not ever updating a row in the database is the safest way to go.
Especially if there is ever a need to recover, this record always exists.

> ## How do we create these two ciphertext blobs?

Reviewer: From context I'm inferring that `ciphertext blobs` refers to the `enc` attribute on the keystore records, but it would be great if you could make it clearer that that's what you're referring to... and maybe explain what the problem is here (i.e. why do we need to give special consideration to how we generate these blobs)

By having encryption context associated with the branch the access of these keys can be controlled by this information.

A point of complexity here is that there are two policy evaluations that are relevant here.
First the evaluation to get access to the branch key and second on how the branch key is used.
In practical terms this is a KMS policy evaluation to unwrap the branch key.
This gets logged in cloudTrail and give visibility to how and where the branch key is being used.

The second is when the branch key is used to encrypt/decrypt.
When encrypting plaintext this is the requirements for the encryption context associated with the plaintext.
This second policy is out of scope for this document
and can be handled as an additional feature in the future.

## What kinds of encryption context conditions should exist for access to a branch key?

### _Only equality_

This is obviously the simplest.
AWS KMS is the party doing the evaluation.

### Local policy evaluation

Having access to the branch key be done by the keystore is too late.
The secret has already been unwrapped.

## How is this encryption context for access to the branch key added to a branch key?

> ### _In the create branch key, add an optional encryption context values_

Reviewer: ```suggestion
### _In the create branch key function, add an optional parameter for encryption context
```
?

This value SHOULD be immutable to the branch key.
Changing these values changes the authorization for previously encrypted messages.

### Also add on new version

This complicates things greatly.
It is more flexible and could still be immutable but merging these two sets could be tricky to reason about.
This is a feature that can be added later.

> ## How should the encryption context for access be serialized into the key store item?

Reviewer: ```suggestion
## How should the encryption context for access be serialized into the keystore item?
```

4. Construct a `GenerateDataKeyWithoutPlaintext` request with the now authenticated additional values
5. ReEncrypt this new version so that a new Active item can be inserted
6. Insert both new items as a transaction

### Get the values from an unauthenticated meta entry

No, having this data be unauthenticated, but then add it to an authenticated process is crazy.

### Just use the existing values on the active entry without authentication

No, this is less dangerous than just using unauthenticated data, but it creates an authentication race. Having a way to inject unauthenticated data into an authenticated process is a bad idea.

### Get the values from an authenticated meta entry

The authentication could be done with decrypt, but this would give decrypt to the create/rotation process.
This could be constrained with policy, but this is not correct by construction.
If a customer does not apply this policy then the process is more privileged.
If it uses the same reEncrypt as above, then what additional value does this provide?
It also creates a new value that is not a branch key that is now floating around.

## How can we let customers control the branch key id?

> ### _Require an encryption context constraint when creating with a defined branch key id_

Reviewer: ```suggestion
### _Require that encryption context MUST also be provided when creating with a defined branch key id_
```

Something like this; as far as I understand it's a "constraint that encryption context must be provided" and I think this makes it clearer...

Since a customer _could_ just use an existing branch key guid.

### An optional branch key id on create

Simplify given the option is possible.
This makes the feature “easy to use” but it is not “hard to misuse”.

### The option is still not safe

If duplicates are created in a single key store there is a problem.
Even GUIDs do not 1000% remove this possibility, however the new structure does.
Even a duplicate GUID would fail in a single key store.
Given this, a very poor id selection would also fail.
This leads us to our second problem, distributed duplicates.
If the entire system is somehow split brain and a duplicate exists on both sides.
Simple GUIDs does not save this case. There needs to be further information.
And the encryption context constraints provide these controls.

Finally, several customers have asked for this.
With at least one taking on implementing this as custom work.
It is better for us to provide the safest API we can and document the sharp edge.
As opposed to trying to not support this or talk about it.

# One-Way Doors

- API changes live forever.
- Changing the database format renders existing keystores invalid

# Security Considerations

> This creates a mechanisms

Reviewer: ```suggestion
This creates a mechanism
```

to enforce access to branch keys.
Explicit encryption context verification
is more secure than implicit encryption context verification.

# Impact/Other considerations

Explaining this subtlety of encryption context

> is going to increase the complexity of our documentation.

Reviewer: Are we working with docs on this change? Not so much the "why" of these decisions, but more the "how" to use them (e.g. providing custom encryption context, the prefix we add to custom values)

Author: yes

[//]: # "Copyright Amazon.com Inc. or its affiliates. All Rights Reserved."
[//]: # "SPDX-License-Identifier: CC-BY-SA-4.0"

# Customers should control where encryption context is stored

# Definitions

## Conventions used in this document

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).

# Background

The current keystore DDB table relies on the branch key id and version being random.

By making the branch key the partition key we can easily group versions together.
By using GUIDs as the branch key ID, e.g. the partition key,
we have a strong control that each branch key is unique.
However, customers want to use their own id for the branch key to simplify binding a branch keys to their customers.
They could take control of creation and implement their own creation process.
As a customer did this it lead them to ask
“how can I ensure that there does not already exist a branch key for this id?”

> This question is fundamental any branch key creation process.

Reviewer 2: ```suggestion
This question is fundamental to any branch key creation process.
```

But different roles have difference accesses.
KMS policy control is only on strict equality so combining all this information make it easy for our development, but more complicated for customers to use.

## How is the encryption context for access propagated to versions?

### *Authenticate the values from the current active key*

We do not want to give the versioning process access to the key, so decrypt is out.
But we already need reEncrypt to create the active entry.
We could reEncrypt the existing entry to authenticate the existing encryption context.
This means performing an additional correctness step, but this process is not in the hot path and does not occur frequently.
If this reEncrypt does not change the encryption context of the new blob, then the scope of policy protecting the branch keys can not be escaped.
This means that additional policy control could be written to constrain this process so that it can only reEncrypt from a version item to an active item and from active item to the same active item.
But this policy is not required to function, it is just an exercise in least privilege.

The idea here would be

1. Get the current item (if no active item exists fail)

> 2. ReEncrypt this item to encryption context identical encryption context

Reviewer 2: What?

Author: ```suggestion
2. ReEncrypt this item to encryption context identical to the existing encryption context.
```
